### PR TITLE
[manufacturing] manuf_cp_ast_test_execution

### LINKED ISSUE #17388: [manufacturing] Test: manuf_cp_ast_test_execution
### Description

Test point: [manuf_cp_ast_test_execution](https://github.com/lowRISC/opentitan/blob/e8f38cd58f738fe5bb3e5b5e3dbcd1ea91d480dc/sw/device/silicon_creator/manuf/data/manuf_testplan.hjson#L75-L93)

### Test Description

Verify that the tester is able to load/exec test sequences into/from SRAM.

An expected pattern during manufacturing involves loading small program sequences into
SRAM, and use them to run higher level operations such as configuring OTP values in the
OTP hardware partition.

- Pre-load OTP with TEST_UNLOCKED lc_state in pre-silicon environments. Advance to
  state in silicon.
- Select CPU JTAG via strap pins.
- Load program into SRAM to execute OTP word write sequences.
- Trigger OTP writes to HW partition (e.g. device ID).
- Perform reset by toggling POR pin.
- Select CPU JTAG via strap pins.
- Verify that OTP write was successful.

### PR DESCRIPTION
**Pull request by pamaury**
Resolves #17388. the test is not complete at the moment due to some unsolved problem when running the OTP code from SRAM, it seems to mess up with the jtag connection, I don't know why yet.

### CODE REVIEW

**Review by timothytrippel — COMMENTED** [review]
LGTM overall! Once you fill in the SRAM program I can approve.

File: sw/device/silicon_creator/manuf/lib/BUILD
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/BUILD:314]: nit: 
```suggestion
    name = "manuf_cp_ast_test_execution_functest",
```

File: sw/device/silicon_creator/manuf/lib/BUILD
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/BUILD:334]: pro-tip, some [bazel shorthand](https://bazel.build/concepts/labels) when the package name matches the target name in the label you can omit the target name:
```suggestion
    test_harness = "//sw/host/tests/manuf/manuf_cp_ast_test_execution",
```

File: sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs
Comment (timothytrippel) [inline sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs:37]: nit: convention is to use the `execute_test!` macro and put the test code a separate function like [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/tests/manuf/manuf_scrap/src/main.rs;drc=f3575255ade33faab01a686dcabdadd80dde2bf9;l=68), this should enable easier code reuse in the future and give a consistent console message at the start of each test

File: sw/device/silicon_creator/manuf/lib/sram_exec_test.c
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/sram_exec_test.c:28]: can you not just use the `otp_ctrl_testutils_dai_read32()` in the `otp_ctrl_testutils` lib? Did you want to avoid bringing in the whole testutils lib for code size?

File: sw/device/silicon_creator/manuf/lib/sram_exec_test.c
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/sram_exec_test.c:56]: could you use the `ottf_console` [lib](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_framework/ottf_console.c) here?

File: sw/device/silicon_creator/manuf/lib/BUILD
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/BUILD:305]: should the uart DIF be a dep too?

File: sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs
Comment (timothytrippel) [inline sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs:36]: no need to setup flow control here, the device side code isn't sending any ujson commands, nor does it turn on flow control itself

File: sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs
Comment (timothytrippel) [inline sw/host/tests/manuf/manuf_cp_ast_test_execution/src/main.rs:38]: nit:
```suggestion
    // Set straps for the CPU TAP and reset.
```

File: sw/host/opentitanlib/src/test_utils/load_sram_program.rs
Comment (timothytrippel) [inline sw/host/opentitanlib/src/test_utils/load_sram_program.rs:37]: nit, for consistency?
```suggestion
    pub stack_pointer: u32,
```

File: sw/device/silicon_creator/manuf/lib/sram_exec_test.c
Comment (pamaury) [reply sw/device/silicon_creator/manuf/lib/sram_exec_test.c:28]: I was doing some tests and trying to make the SRAM binary small because loading a big binary was causing problems. This will be fixed in the final version.

Comment (timothytrippel) [reply sw/device/silicon_creator/manuf/lib/sram_exec_test.c]: ahh makes sense, yeah that doesn't make sense to be static, you can change it.

File: sw/device/silicon_creator/manuf/lib/sram_exec_test.c
Comment (timothytrippel) [inline sw/device/silicon_creator/manuf/lib/sram_exec_test.c:66]: Is there a reason for the "42" return code?

**Comment by timothytrippel** [comment]
I think one of your CI errors (in addition to your lint failures) is that you need a header file for your SRAM program that has an entry point alias like here: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/examples/sram_program/sram_program.h

**Comment by timothytrippel** [comment]
As we discussed earlier today, I pulled this down and ran it several times. When I commented out
1. the application of the RMA_BOOTSTRAP strapping in the host-side rust program, and
2. everything on the device side C program except the setting of the pinmux, UART console config, and the `LOG_INFO("Hello from SRAM");`, and
switched to the bitstream that has the OTP image where ROM execution is disabled, I get consistent test passing on my machine / FPGA (actually I wasn't able to produce a failure):

Screenshot 2023-05-08 at 10 38 05 PM

There is something not quite right about the OTP reads, but I can look into that more tomorrow if you still can't get that part working.

File: sw/device/silicon_creator/manuf/lib/sram_exec_test.c
Comment (pamaury) [reply sw/device/silicon_creator/manuf/lib/sram_exec_test.c:66]: Not it's just to debugging purposes, this will go away when the test is done properly.

**Review by timothytrippel — APPROVED** [review]
I spoke with @pamaury offline, we are going to get this merged as is, and I will clean up / get OTP interactions working in a follow up PR.

**Pull request merged**
